.gitlab-ci.yml

```diff
@@ -11,6 +15,59 @@ include:
   - project: 'Northern.tech/Mender/mendertesting'
     file: '.gitlab-ci-github-status-updates.yml'
 
+.set_eks_helmci_vars: &set_eks_helmci_vars
+  - export AWS_ACCESS_KEY_ID=$CI_EKSCTL_ACCESS_KEY_ID_TEST
+  - export AWS_SECRET_ACCESS_KEY=$CI_EKSCTL_AWS_SECRET_ACCESS_KEY_TEST
+  - export AWS_DEFAULT_REGION=eu-central-1
+
+build:pipeline-toolbox-image:
+  tags:
+    - mender-qa-worker-generic-light
+  stage: .pre
+  image: docker
+  services:
+    - docker:dind
+  rules:
+    - if: '$CI_COMMIT_BRANCH == "master" || $CI_COMMIT_BRANCH == "master-next" || $RUN_HELM_CHART_INSTALL == "true" || $CI_COMMIT_TAG'
+  script:
+    - echo $CI_REGISTRY_PASSWORD | docker login -u $CI_REGISTRY_USER $CI_REGISTRY --password-stdin
+    - |
+      docker build \
+            --build-arg BUILDKIT_INLINE_CACHE=1 \
+            --cache-from ${CI_REGISTRY_IMAGE}:${PIPELINE_TOOLBOX_TAG} \
+            -t ${CI_REGISTRY_IMAGE}:${PIPELINE_TOOLBOX_TAG} \
+            --build-arg KUBECTL_VERSION=${CI_KUBECTL_VERSION} \
+            --build-arg HELM_VERSION=${CI_HELM_VERSION} \
+            --build-arg KUBECONFORM_VERSION=${CI_KUBECONFORM_VERSION} \
+            -f tests/Dockerfile \
+            tests/
+    - docker push ${CI_REGISTRY_IMAGE}:${PIPELINE_TOOLBOX_TAG}
```

Comment:
The best way is to have a global image, used also in other pipelines. This requires a dedicated repo and dedicated publishing pipeline. Since this a huge move, I'm taking the small step instead.

Here the build time is mitigated by the cache: `--cache-from ${CI_REGISTRY_IMAGE}:${PIPELINE_TOOLBOX_TAG}` and it's really fast (~1min). 
Having this at the beginning of the pipeline gives the opportunity to modify the Dockerfile as needed and simply launch a new pipeline.

The alternative is to use the base image `debian:slim` and building the tools at every step, but it seems worst to me...

Keep in mind that this image will be later reused for Helm version bump automation made with the `yq` tool.

Reply:
I don't have a problem with building and publishing the image, it's just that it's not necessary to do for every run.  I'm mainly concerned about the overhead of pushing the same image for every run. Also, if the registry has immutable tags, the push command will fail.
Can we make this a manual build step for simplicity?

Reply:
Yep, it's better. Done

Reply:
What about building it only when the `Dockerfile` was changed?

```
  rules:
    - if: '$CI_COMMIT_BRANCH == "master" || $CI_COMMIT_BRANCH == "master-next" || $RUN_HELM_CHART_INSTALL == "true" || $CI_COMMIT_TAG'
      changes:
        - tests/Dockerfile
```

Reply:
Afaik that doesn't work due to the mirroring we do for the gitlab connection... But using a pre-built image is what we do in other areas with mender-test-containers - so it might not be a lot of extra effort to set up...

Reply:
Interesting.. Actually I can't find it in the `.gitlab-ci.yml` of the `mender-test-containers`.. Can you please spot me it out?

About the gitlab mirroring, I don't get it: the `mender-test-runner` is pushing a new commit with the modified `Dockerfile` to Gitlab, then the Gitlab Pipeline, when merged to `master`, should run the toolbox image build.. If no changes from the previous commit, then no toolbox image build. For sure I'm missing something, though.

Anyway, since this an implementation detail, let's set it as a manual step, will improve it along the way ;). I re-set it to `when: manual` as suggested by @alfrunes

Reply:
e.g. https://github.com/mendersoftware/mender-test-containers/blob/master/.gitlab-ci.yml#L65-L80 builds a testing/ tools image for the gui e2e tests - that's not run on change, but at least far less frequent than the pipeline itself...

As to the mirroring problems, if I remember correctly the change detection is made in a merge request, whereas we force push the entire branch and Gitlab just runs CI with the latest commit - not caring "how it got to that commit"...

I may be wrong so feel free to try it out, just as a word of caution that we unfortunately can't rely on all the Gitlab functionalities...

Reply:
> e.g. https://github.com/mendersoftware/mender-test-containers/blob/master/.gitlab-ci.yml#L65-L80 builds a testing/ tools image for the gui e2e tests - that's not run on change, but at least far less frequent than the pipeline itself...
> 
> As to the mirroring problems, if I remember correctly the change detection is made in a merge request, whereas we force push the entire branch and Gitlab just runs CI with the latest commit - not caring "how it got to that commit"...
> 
> I may be wrong so feel free to try it out, just as a word of caution that we unfortunately can't rely on all the Gitlab functionalities...

I can confirm that the change detection is not working on our setup. Thanks for the heads-up

Reply:
AFAIK it works as described in [the docs](https://docs.gitlab.com/ee/ci/yaml/#ruleschanges) (which is not to say that it works as I intuitively expect). Basically for every new branch and new PR it will always run. Plus on scheduled pipelines, tags, manual runs etc.

I suggest we use `mender-test-containers` for the test image as Manuel is suggesting. This is a good way to align the base test images across all our CI.

Reply:
Solved by using a `mender-test-containers` custom pipeline. See: https://github.com/mendersoftware/mender-test-containers/pull/342